lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs

```diff
@@ -106,84 +110,103 @@ random
     -> NonEmpty TxOut
     -> UTxO
     -> ExceptT ErrCoinSelection m (CoinSelection, UTxO)
-random opt outs utxo = do
+random opt outs utxo0 = do
     let descending = NE.toList . NE.sortBy (flip $ comparing coin)
     randomMaybe <- lift $ runMaybeT $ foldM
-        (processTxOut opt)
-        (utxo, mempty)
+        (makeSelection opt)
+        (utxo0, [])
         (descending outs)
     case randomMaybe of
-        Just (utxo', res) ->
-            return (res, utxo')
+        Just (utxo1, res) -> do
+            (utxo2, selection) <-
+                lift $ foldM
+                (improveTxOut opt)
+                (utxo1, mempty)
+                (reverse res)
+            return (selection, utxo2)
         Nothing ->
-            largestFirst opt outs utxo
+            largestFirst opt outs utxo0
 
--- | Perform a random selection on a given output, with improvement.
-processTxOut
+-- | Perform a random selection on a given output, without improvement.
+makeSelection
     :: forall m. MonadRandom m
     => CoinSelectionOptions
-    -> (UTxO, CoinSelection)
+    -> (UTxO, [([(TxIn, TxOut)], TxOut)])
     -> TxOut
-    -> MaybeT m (UTxO, CoinSelection)
-processTxOut (CoinSelectionOptions maxNumInputs) (utxo0, selection) txout = do
-    attempt <- coverRandomly ([], utxo0)
-    (inps, utxo') <- lift (improve attempt)
+    -> MaybeT m (UTxO, [([(TxIn, TxOut)], TxOut)])
```

Comment:
Why did we get rid of the `CoinSelection` type here? Because we do this in two steps, the type shouldn't be much different since the second step is somewhat similar to `(UTxO, CoinSelection) -> (UTxO, CoinSelection)`.

It becomes very noisy quickly with all the `TxIn` and `TxOut` sums and products. So if nothing justifies it strongly, I'd rather stay with a simpler type-signature :+1:

Reply:
in the first phase, we are constructing selections for all outputs and the utxo that is diminished by doing so. basically, in the phase two I start for each output with the selected inputs and try to improve with current utxo. only then I construct coinselection (where there is also change computed). So, for me, it was easier to operate in phase one and as an input to phase two with `(inputs, txout)` for each output.

Reply:
I see one obstacle to use CoinSelection at phase 1. If I use `<>` when processing outputs, I will coalesce all inputs from different outputs. We have :  
```
72 data CoinSelection = CoinSelection                                                                                                                                                         
73     { inputs  :: [(TxIn, TxOut)]                                                                                                                                                           
74       -- ^ Picked inputs                                                                                                                                                                   
75     , outputs :: [TxOut]                                                                                                                                                                   
76       -- ^ Picked outputs                                                                                                                                                                  
77     , change  :: [Coin]                                                                                                                                                                    
78       -- ^ Resulting changes                                                                                                                                                               
79     } deriving (Generic, Show, Eq) 
```
And then I will not have this knowledge at stage 2 (when doing improvement). And here I need exact inputs per output. So using coin selection at phase 1 somewhat blur boundaries upon `<>` Maybe I am missing something...

Reply:
Nevertheless, `[([(TxIn, TxOut)], TxOut)]` may look cryptic, although this represents the list of `(inps, txout)` ie., list of all inps selected for a given txout

Reply:
:thinking: ... hmmm. Since we treat each output independently, would it make sense to actually consider `[CoinSelection]` at this stage, with "singleton" selections? This way, we preserve the relationship between inputs/outputs and we still get a rather readable type ?

Reply:
Yeah. I am just grooming hoping to find something slightly better, but it's not like it's totally cryptic. In the end, as you said, it's quite natural and understandable from the context.